apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nav-sidebar/src/components/nav-sidebar/index.tsx

```diff
@@ -33,10 +34,13 @@ const Button = forwardRef(
 		{
 			children,
 			...rest
-		}: OriginalButton.Props & { icon?: any; iconSize?: number; showTooltip?: boolean },
+		}: OriginalButton.Props & { icon?: unknown; iconSize?: number; showTooltip?: boolean },
 		ref
 	) => (
-		<OriginalButton ref={ ref as any } { ...rest }>
+		<OriginalButton
+			ref={ ref as string | ( ( instance: HTMLElement | null ) => void ) }
```

Comment:
`ref` types were weird here; basically just copy pasted from the type error.

Reply:
Should it be `as React.Ref<HTMLElement>`?

Reply:
That'd be nice, but it gives an error:

```
No overload matches this call.
  Overload 1 of 2, '(props: Props, context?: any): ReactElement<any, any> | Component<Props, any, any> | null', gave the following error.
    Type 'Ref<HTMLElement>' is not assignable to type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null | undefined'.
      Type 'RefObject<HTMLElement>' is not assignable to type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null | undefined'.
        Type 'RefObject<HTMLElement>' is not assignable to type 'RefObject<HTMLAnchorElement>'.
          Type 'HTMLElement' is missing the following properties from type 'HTMLAnchorElement': charset, coords, download, hreflang, and 21 more.
  Overload 2 of 2, '(props: PropsWithChildren<Props>, context?: any): ReactElement<any, any> | Component<Props, any, any> | null', gave the following error.
    Type 'Ref<HTMLElement>' is not assignable to type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null | undefined'.
      Type 'RefObject<HTMLElement>' is not assignable to type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null | undefined'.
        Type 'RefObject<HTMLElement>' is not assignable to type 'RefObject<HTMLAnchorElement>'.
```

I think the types on `Button` from @wordpress/components must be missing something 🤔

Reply:
Ah, no it's because `Button` can render either an `HTMLButtonElement` or a `HTMLAnchorElement`.

`HTMLElement` isn't a super type of `HTMLButtonElement` or `HTMLAnchorElement` so neither can fit into that box. You'll want to change it to `Ref<HTMLAnchorElement | HTMLButtonElement>` I think (though try just `HTMLAnchorElement` first?)

Reply:
It still doesn't work with button, anchor, or button/anchor together :/

```
No overload matches this call.
  Overload 1 of 2, '(props: Props, context?: any): ReactElement<any, any> | Component<Props, any, any> | null', gave the following error.
    Type '{ children: ReactNode; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; isDestructive?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; } | { ...; }' is not assignable to type '(IntrinsicAttributes & AnchorProps) | (IntrinsicAttributes & ButtonProps)'.
      Type '{ children: ReactNode; href?: undefined; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; }' is not assignable to type '(IntrinsicAttributes & AnchorProps) | (IntrinsicAttributes & ButtonProps)'.
        Type '{ children: ReactNode; href?: undefined; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; }' is not assignable to type 'ButtonProps'.
          Types of property 'ref' are incompatible.
            Type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
              Type '(instance: HTMLAnchorElement | null) => void' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
                Type '(instance: HTMLAnchorElement | null) => void' is not assignable to type '(instance: HTMLButtonElement | null) => void'.
                  Types of parameters 'instance' and 'instance' are incompatible.
                    Type 'HTMLButtonElement | null' is not assignable to type 'HTMLAnchorElement | null'.
                      Type 'HTMLButtonElement' is missing the following properties from type 'HTMLAnchorElement': charset, coords, download, hreflang, and 19 more.
  Overload 2 of 2, '(props: PropsWithChildren<Props>, context?: any): ReactElement<any, any> | Component<Props, any, any> | null', gave the following error.
    Type '{ children: ReactNode; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; isDestructive?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; } | { ...; }' is not assignable to type '(IntrinsicAttributes & AnchorProps & { children?: ReactNode; }) | (IntrinsicAttributes & ButtonProps & { ...; })'.
      Type '{ children: ReactNode; href?: undefined; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; }' is not assignable to type '(IntrinsicAttributes & AnchorProps & { children?: ReactNode; }) | (IntrinsicAttributes & ButtonProps & { ...; })'.
        Type '{ children: ReactNode; href?: undefined; isDefault?: boolean | undefined; isPrimary?: boolean | undefined; isSecondary?: boolean | undefined; isTertiary?: boolean | undefined; ... 371 more ...; key?: string | ... 2 more ... | undefined; }' is not assignable to type 'ButtonProps'.
          Types of property 'ref' are incompatible.
            Type 'string | ((instance: HTMLAnchorElement | null) => void) | RefObject<HTMLAnchorElement> | ((instance: HTMLButtonElement | null) => void) | RefObject<...> | null' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
              Type '(instance: HTMLAnchorElement | null) => void' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
                Type '(instance: HTMLAnchorElement | null) => void' is not assignable to type '(instance: HTMLButtonElement | null) => void'.
```

Reply:
I tried a few other things when I was working through it, and so far the only thing that clears it up is the code I committed 🙃

Reply:
That's really unfortunate...

For reference, here's the DefinitelyTyped definition for `Button`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/button/index.d.ts

I wonder what needs to be fixed there to make this work appropriately? The ref type doesn't seem to be documented at all. I wonder if it should be `ForwardRefExoticComponent` with the appropriate ref attributes put in instead of just `ComponentType`? 😓 What a mess 😕

Reply:
oof. I wonder if this is something we could add more generally (e.g. it seems like most components should accept ref/classname)

Reply:
It'll be easier when we move strictly to JSDoc/TypeScript for components. Not sure what to do about the DefinitelyTyped declarations though.

You're right that most components should forward refs, unfortunately many do not so it's still a work in progress. G2 components all forward refs though, so that's good!